File: vuln-sandbox/python/cwe-78/command-injection.py
# VULNERABLE: InjectPythonSubprocessShell — user input passed to subprocess with shell=True
# Rule: InjectPythonSubprocessShellTemplate | CWE-78 | Severity: CRITICAL

import subprocess
from flask import Flask, request, jsonify

app = Flask(__name__)


@app.route('/api/ping', methods=['POST'])
def ping_host():
    """Ping a host provided by the user."""
    data = request.get_json()
    host = data.get('host', '')

    # <-- VULNERABLE: shell=True with user-controlled cmd enables arbitrary command execution
    # An attacker can pass: "8.8.8.8; rm -rf /" to run arbitrary shell commands
    cmd = f"ping -c 1 {host}"
    result = subprocess.run(cmd, shell=True, capture_output=True, text=True)

    return jsonify({
        'stdout': result.stdout,
        'stderr': result.stderr,
        'returncode': result.returncode,
    })


if __name__ == '__main__':
    app.run()

File: vuln-sandbox/node/cwe-78/command-injection.js
// VULNERABLE: command-injection — spawn with shell:true and user-controlled command
// Rule: InjectChildProcessShellTrue | CWE-78 | Severity: CRITICAL

const express = require('express');
const { spawn } = require('child_process');

const app = express();
app.use(express.json());

app.post('/run', (req, res) => {
  const cmd = req.body.command;
  const args = req.body.args || [];

  // VULNERABLE: shell:true allows shell metacharacter injection
  const proc = spawn(cmd, args, { shell: true });

  let output = '';
  proc.stdout.on('data', (data) => { output += data.toString(); });
  proc.stderr.on('data', (data) => { output += data.toString(); });
  proc.on('close', (code) => {
    res.json({ output, exitCode: code });
  });
});

app.listen(3000);
